torchtext/datasets/iwslt2016.py

```diff
@@ -225,50 +205,62 @@ def IWSLT2016(root='.data', split=('train', 'valid', 'test'), language_pair=('de
-    # We need to take an extra step to pick out the specific language pair from it.
+    url_dp = IterableWrapper([URL])
+    cache_compressed_dp = url_dp.on_disk_cache(
+        filepath_fn=lambda x: os.path.join(root, _PATH),
+        hash_dict={os.path.join(root, _PATH): MD5},
+        hash_type="md5"
+    )
+    cache_compressed_dp = GDriveReader(cache_compressed_dp).end_caching(mode="wb", same_filepath_fn=True)
+    cache_compressed_dp = FileOpener(cache_compressed_dp, mode="b")
     src_language = train_filenames[0].split(".")[-1]
     tgt_language = train_filenames[1].split(".")[-1]
     languages = "-".join([src_language, tgt_language])
 
-    iwslt_tar = '{}/{}/texts/{}/{}/{}.tgz'
-    iwslt_tar = iwslt_tar.format(
-        root, SUPPORTED_DATASETS['_PATH'].split(".")[0], src_language, tgt_language, languages)
-    extracted_dataset_tar = extract_archive(iwslt_tar)
-    extracted_files.extend(extracted_dataset_tar)
-
-    # Clean the xml and tag file in the archives
-    file_archives = []
-    for fname in extracted_files:
+    iwslt_tar = os.path.join(
+        "texts", src_language, tgt_language, languages
+    )
+    cache_decompressed_dp = cache_compressed_dp.on_disk_cache(
+        # Convert /path/to/downloaded/foo.tgz to /path/to/downloaded/foo/rest/of/path
+        filepath_fn=lambda x: os.path.join(os.path.splitext(x[0])[0], iwslt_tar)
+    )
+    cache_decompressed_dp = cache_decompressed_dp.read_from_tar()
+    cache_decompressed_dp = cache_decompressed_dp.end_caching(mode="wb")
+
+    def clean_files(fname):
         if 'xml' in fname:
             _clean_xml_file(fname)
-            file_archives.append(os.path.splitext(fname)[0])
+            return os.path.splitext(fname)[0]
         elif "tags" in fname:
             _clean_tags_file(fname)
-            file_archives.append(fname.replace('.tags', ''))
-        else:
-            file_archives.append(fname)
-
-    data_filenames = {
-        "train": _construct_filepaths(file_archives, src_train, tgt_train),
-        "valid": _construct_filepaths(file_archives, src_eval, tgt_eval),
-        "test": _construct_filepaths(file_archives, src_test, tgt_test)
-    }
-
-    for key in data_filenames.keys():
-        if len(data_filenames[key]) == 0 or data_filenames[key] is None:
-            raise FileNotFoundError(
-                "Files are not found for data type {}".format(key))
-
-    src_data_iter = _read_text_iterator(data_filenames[split][0])
-    tgt_data_iter = _read_text_iterator(data_filenames[split][1])
-
-    def _iter(src_data_iter, tgt_data_iter):
-        for item in zip(src_data_iter, tgt_data_iter):
-            yield item
-
-    return _RawTextIterableDataset(DATASET_NAME, NUM_LINES[split][num_lines_set_identifier[split]][tuple(sorted(language_pair))], _iter(src_data_iter, tgt_data_iter))
+            return fname.replace('.tags', '')
+        return fname
+
+    cache_decompressed_dp = cache_decompressed_dp.flatmap(FileLister)
```

Comment:
So I guess this is doing the magic for flattening out all the inner files after extracting from inner tarballs right? I am trying to infer the role of FileLister based on your comment here  https://github.com/pytorch/text/issues/1494#issuecomment-1019716382 . It's not so intuitive that FileLister is doing the extraction. Could you please help understand a bit what this line is doing?

Reply:
Yes, `flatmap(FileLister)` is not doing the extraction but is instead effectively linearizing the file tree of the directory structures which are the output of the `read_from_tar`.

So basically `read_from_tar()` will yield something like:

```
(/path/to/dir1/, /path/to/dir2/)`
```

And flatmapping the `FileLister` basically walks those trees in order.

Does this make sense?

Reply:
But we are calling `read_from_tar` only once here (https://github.com/pytorch/text/pull/1545/files#diff-6ea166aeffab6173bf513a4028091fcc22977b25d4b54fc98026ce0a0ad44476R227), which shall extract files from `2016-01.tgz`. I am trying to understand where do we extract from inner tarballs, for eg: for language pair (de-en)we would need to extract from inner tarball `de-en.tgz` . Sorry i may be missing something obvious here?

Reply:
No, you're not missing anything obvious. I think the code is not doing what I expect and I'm trying to figure it out. The issue seems to be about `filepath_fn` not doing what I think it should do since it's used to extract files, but not directories...